crates/bevy_pbr/src/ssao/preprocess_depth.wgsl

```diff
@@ -0,0 +1,104 @@
+@group(0) @binding(4) var preprocessed_depth_mip3: texture_storage_2d<r16float, write>;
+@group(0) @binding(5) var preprocessed_depth_mip4: texture_storage_2d<r16float, write>;
+@group(1) @binding(0) var point_clamp_sampler: sampler;
+@group(1) @binding(1) var<uniform> view: View;
+
+
+// Using 4 depths from the previous MIP, compute a weighted average for the depth of the current MIP
+fn weighted_average(depth0: f32, depth1: f32, depth2: f32, depth3: f32) -> f32 {
+    // TODO: Cleanup constants
+    // TODO: Document how the weights are determined, and what the parameters are doing
+    let depth_range_scale_factor = 0.75;
+    let effect_radius = depth_range_scale_factor * 0.5 * 1.457;
+    let falloff_range = 0.615 * effect_radius;
+    let falloff_from = effect_radius * (1.0 - 0.615);
+    let falloff_mul = -1.0 / falloff_range;
+    let falloff_add = falloff_from / falloff_range + 1.0;
+    let max_depth = max(max(depth0, depth1), max(depth2, depth3));
+
+    let weight0 = saturate((max_depth - depth0) * falloff_mul + falloff_add);
+    let weight1 = saturate((max_depth - depth1) * falloff_mul + falloff_add);
+    let weight2 = saturate((max_depth - depth2) * falloff_mul + falloff_add);
+    let weight3 = saturate((max_depth - depth3) * falloff_mul + falloff_add);
+    let weight_total = weight0 + weight1 + weight2 + weight3;
+
+    let depth = ((weight0 * depth0) + (weight1 * depth1) + (weight2 * depth2) + (weight3 * depth3)) / weight_total;
+    return min(depth, 3.402823466e+38); // Clamp INF to f32::MAX
+}
+
+// Used to share the depths from the previous MIP level between all invocations in a workgroup
+var<workgroup> previous_mip_depth: array<array<f32, 8>, 8>;
+
+@compute
+@workgroup_size(8, 8, 1)
+fn preprocess_depth(@builtin(global_invocation_id) global_id: vec3<u32>, @builtin(local_invocation_id) local_id: vec3<u32>) {
+    let base_coordinates = vec2<i32>(global_id.xy);
+
+    // MIP 0 - Copy 4 texels from the input depth (per invocation, 8x8 invocations per workgroup)
+    let pixel_coordinates0 = base_coordinates * 2i;
+    let pixel_coordinates1 = pixel_coordinates0 + vec2<i32>(1i, 0i);
+    let pixel_coordinates2 = pixel_coordinates0 + vec2<i32>(0i, 1i);
+    let pixel_coordinates3 = pixel_coordinates0 + vec2<i32>(1i, 1i);
+    let depths_uv = vec2<f32>(pixel_coordinates0) / view.viewport.zw;
+    let depths = textureGather(0, input_depth, point_clamp_sampler, depths_uv, vec2<i32>(1i, 1i));
+    textureStore(preprocessed_depth_mip0, pixel_coordinates0, vec4<f32>(depths.w, 0.0, 0.0, 0.0));
+    textureStore(preprocessed_depth_mip0, pixel_coordinates1, vec4<f32>(depths.z, 0.0, 0.0, 0.0));
+    textureStore(preprocessed_depth_mip0, pixel_coordinates2, vec4<f32>(depths.x, 0.0, 0.0, 0.0));
+    textureStore(preprocessed_depth_mip0, pixel_coordinates3, vec4<f32>(depths.y, 0.0, 0.0, 0.0));
+
+    // MIP 1 - Weighted average of MIP 0's depth values (per invocation, 8x8 invocations per workgroup)
```

Comment:
I think my math around the comments of how many invocations we have per workgroup is wrong...

Reply:
Nvm